Improve Contribution Doc (#2043)
huggingface/diffusers: docs/source/en/conceptual/contribution.mdx

Everyone is encouraged to start by saying 👋 in our public Discord channel. We discuss the hottest trends about diffusion models, ask questions, show-off personal projects, help each other with contributions, or just hang out ☕. <a href="https://discord.gg/G7tWnz98XR"><img alt="Join us on Discord" src="https://img.shields.io/discord/823813159592001537?color=5865F2&logo=discord&logoColor=white"></a>

Whichever way you choose to contribute, we strive to be part of an open, welcoming and kind community. Please, read our [code of conduct](https://github.com/huggingface/diffusers/blob/main/CODE_OF_CONDUCT.md) and be mindful to respect it during your interactions.

That being said, we enourmously value feedback from the community, so please do not be afraid to speak up if you believe you have valueable feedback that can help improve the library - every message, comment, issue and pull request (PR) is read and considered.

## Overview

You can contribute in many ways ranging from answering questions on issues to adding new diffusion models to 
the core library.

> In the following, we give an overview of different ways to contribute, ranked by difficulty in ascending order.

Reviewer: Actually, I liked the encouraging sentence that was deleted below :) Maybe consider putting it back as part of this introduction?

```suggestion
In the following, we give an overview of different ways to contribute, ranked by difficulty in ascending order. All of them are valuable to the community.
```

* 1. Asking and answering questions on [the diffusers discussion forum](TODO:PVP) or on [discord](TODO:PVP).
* 2. Opening new issues on [the GitHub Issues tab](TODO:PVP)
* 3. Answering issues on [the GitHub Issues tab](TODO:PVP)
* 4. Fix a simple issue, marked by the "Good first issue" label, see [here](TODO:PVP).
* 5. Contribute to the [documentation](https://github.com/huggingface/diffusers/tree/main/docs/source).
* 6. Contribute a [Community Pipeline](TODO:PVP)
* 7. Contribute to the [examples](https://github.com/huggingface/diffusers/tree/main/examples).
* 8. Fix a more difficult issue, marked by the "Good second issue" label, see [here](TODO:PVP).
* 9. Add a new pipeline, model and, or pipeline, see ["New Pipeline/Model"](TODO:PVP) and ["New scheduler"](TODO:PVP) issues. For this contribution, please have a look at [Design Philosophy](TODO:PVP).

> As said before, **all contributions are valuable to the community**.

Reviewer: Remove "As said before" if we do remove the previous mention.

In the following, we will explain each contribution a bit more in-detail.

> For all contributions 4.-9. you will need to open a PR. It is explained in detail how to do so in [Opening a pull requst](TODO:PVP)

Reviewer: ```suggestion
For all contributions 4.-9. you will need to open a PR. It is explained in detail how to do so in [Opening a pull request](TODO:PVP)
```

### 1. Asking and answering questions on [the diffusers discussion forum](TODO:PVP)

	The diffusers community is incentivized to ask any question related to the library and/or personal 
	projects built on top of the library on the `diffusers` [discussion forum](TODO:PVP) or the `diffusers` [discord channels](TODO:PVP) in an attempt to share 
	knowledge, receive help, network, or just to present personal projects. 

> 	Every questions that is asked on the forum or on discord, actively encourages the community to publicly 

Reviewer: ```suggestion
	Every question that is asked on the forum or on discord, actively encourages the community to publicly 
```

> 	share knowledge and might very well help a beginner in the future that has the same question, you're 
> 	having. Please do pose any question you might have.

Reviewer: ```suggestion
	share knowledge and might very well help a beginner in the future that has the same question you're 
	having now. Please, do post any question you might have.
```

	is the quality of the publicly documented knowledge. The same way, well-posed and well-answered questions    create a high quality knowledge database accesible to everybody, a badly posed question or answer reduces the overall quality of the public knowledge database. 
	In short, a high quality question or answer is *precise*, *concise*, *relevant*, *easy-to-understand*, *accesible*, and *well-formated/well-posed*. For more information, please have a look through the [How to write a good issue](#how-to-write-a-good-issue) section.

### 2. Opening new issues on the GitHub issues tab

	The 🧨 Diffusers library is robust and reliable thanks to the users who notify us of
	the problems they encounter. So thank you for reporting an issue.

    Remember, GitHub issues are reserved for technical questions directly related to the `diffusers` library, 	bug reports, feature requests, or feedback on the library design.

	In a nutshell, this means that everything that is **not** related to the **code of the diffusers library** should **not** be asked on GitHub, but rather on either the [forum](TODO:PVP) or [discord](TODO:PVP) as explained in [1.](TODO:PVP).

> 	**Please consider the following guidelines when open a new issue**:

Reviewer: ```suggestion
	**Please consider the following guidelines when you open a new issue**:
```

	- Make sure you have searched whether your issue has already been asked before (use the search bar on GitHub under Issues).
	- Please never report a new issue on another (related) issue. If another issue is highly related, please
	open a new issue nevertheless and link to the related issue.

> 	- Make sure your issue is written in English. Plesae use on of the great, free online trasnlation services, such as [DeepL](TODO:PVP) to translate from your native language to English if you are not conformtible in English.

Reviewer: ```suggestion
	- Make sure your issue is written in English. Plesae use one of the great, free online trasnlation services, such as [DeepL](TODO:PVP) to translate from your native language to English if you are not comfortable in English.
```

> 	- Check whether your issue might be solved by updating to the newest `diffusers` version. Before posting your issue, please make sure that `python -c "import diffusers; print(diffusers.__version__)"` is higher or matches the latest `diffusers` version.

Reviewer: ```suggestion
	- Check whether your issue might be solved by updating to the newest `diffusers` version. Before posting your issue, please make sure that `python -c "import diffusers; print(diffusers.__version__)"` is higher (when working on the `main` git branch) or matches the latest `diffusers` version.
```

	- Remember that the more effort you put into opening a new issue, the higher the quality of your answer will be and the better the overall quality of the `diffusers` issues.

New issues usually include the following.

#### 2.1. Reproducible, minimal bug reports.

A bug report should always have a reproducible code snippet and be as minimal and concise as possible.
This means in more detail:
- Narrow the bug down as much as you can, **do not just dump your whole code file**
- Format your code

> - Do not include any external libraries except for `diffusers` depends on them.

Reviewer: I'm not sure what you mean here.

- **Always** provide all necessary information about your environment; for this you can run: `diffusers-cli env` in your shell and copy-paste the displayed information to the issue.
- Explain the issue. If the reader doesn't know what the issue is and why it is an issue, she cannot solve it.

> - **Always** make sure the reader can reproduce your issue with as little effort as possible. If you code snippet cannot be run because of missing libraries or undefined variables, the reader cannot help you.  Make sure your reproducible code snippet is as minimal as possible and can be copy-pasted into a simple Python shell.

Reviewer: ```suggestion
- **Always** make sure the reader can reproduce your issue with as little effort as possible. If your code snippet cannot be run because of missing libraries or undefined variables, the reader cannot help you.  Make sure your reproducible code snippet is as minimal as possible and can be copy-pasted into a simple Python shell.
```

> - If in order to reproduce your issue a model and/or dataset is required make sure the reader has access to the model or dataset. You can always upload your model or dataset to the [Hub](https://huggingface.co) to make it easily downloadable. Try to keep your model and dataset as small as possible, to make the reproduction of your issue as effortless as possible.

Reviewer: ```suggestion
- If in order to reproduce your issue a model and/or dataset is required, make sure the reader has access to that model or dataset. You can always upload your model or dataset to the [Hub](https://huggingface.co) to make it easily downloadable. Try to keep your model and dataset as small as possible, to make the reproduction of your issue as effortless as possible.
```

4. In case this is related to a paper, please attach a link;
5. Attach any additional information (drawings, screenshots, etc.) you think may help.

You can open a feature request [here](TODO:PVP).

#### 2.3 Feedback. 

> Feedback about the library design and why it is good or not good helps the core maintainers immensely to build a user-friendly library. To understand the philosohpy behind the current design philosophy, plesae have a look [here](TODO:PVP). If you feel like a certain design choice does not fit in with the current design philosophy, plesae explain why and how it should be changed. If a certain design choice follows the design philosophy too much, hence restricting use cases, explain why and how it should be changed. 

Reviewer: ```suggestion
Feedback about the library design and why it is good or not good helps the core maintainers immensely to build a user-friendly library. To understand the philosohpy behind the current design philosophy, please have a look [here](TODO:PVP). If you feel like a certain design choice does not fit in with the current design philosophy, please explain why and how it should be changed. If a certain design choice follows the design philosophy too much, hence restricting use cases, explain why and how it should be changed. 
```

> If a certain design choice is very useful for you, please also leave a not as this is great feedback for future design decisions.

Reviewer: ```suggestion
If a certain design choice is very useful for you, please also leave a note as this is great feedback for future design decisions.
```

You can open an issue about a feedback [here](TODO:PVP).

#### 2.4 Technical questions. 

> Technical questions are mainly about why certain code of the library was written in a certain way, or what a certain part of the code does. Please make sure to link to the code in question and plesae provide detail on

Reviewer: ```suggestion
Technical questions are mainly about why certain code of the library was written in a certain way, or what a certain part of the code does. Please make sure to link to the code in question and please provide detail on
```

why this part of the code is difficult to understand.

You can open an issue about a technical question [here](TODO:PVP).

#### 2.5 Proposal to add a new model, scheduler, or pipeline.

> If the diffusion model community relesaed a new model, pipeline or scheduler that you would like to see in the `diffusers` library, please provide the following information:

Reviewer: ```suggestion
If the diffusion model community released a new model, pipeline or scheduler that you would like to see in the `diffusers` library, please provide the following information:
```

* Short description of the diffusion pipeline, model or scheduler and link to the paper or public release.
* Link to the implementation if it is open-source;.
* Link to the model weights if they are available.

If you are willing to contribute the model yourself, let us know so we can best guide you. Also don't forget 
to tag the author by github handle if you can find it.

> You can open a requests for a model/pipeline [here](TODO:PVP) and for a scheduler [here](TODO:PVP).

Reviewer: ```suggestion
You can open requests for a model/pipeline [here](TODO:PVP) and for a scheduler [here](TODO:PVP).
```

	- a.) The issue description already proposes a fix. In this case and if the solution makes sense to you, you can open a PR or work-in-progress PR to fix it. 
	- b.) The issue description does not propose a fix. In this case, you can ask how a proposed fix could look like and someone should answer shortly. If you have a good idea on how to fix it, feel free to directly open a PR.
	- c.) There is already on open PR to fix the issue, but the issue hasn't been closed yet. If the PR has gone stale, you can simply open a new PR and link to the stale PR. PRs often go stale if the original contributor who wanted to fix the issue suddently cannot find the time anymore to proceed. This often happens in open-source and is very normal. In this case, the community will be very happy if you give it a new try and leverage knowledge of the existing PR. If there is already a PR and it is active, you can help the author by giving suggestions, reviewing the PR or even asking whether you can contribute to the PR.

### 5. Contribute to the [documentation](https://github.com/huggingface/diffusers/tree/main/docs/source).

> 	A good library **always** has good documentation! The official documentation is often one of the first point of contacts with a user of the library and therefore contributing to the documentation is a **highly 

Reviewer: ```suggestion
	A good library **always** has good documentation! The official documentation is often one of the first points of contact for new users of the library, and therefore contributing to the documentation is a **highly 
```

	valuable contribution**.

> 	Contributing to the library can have many forms:

Reviewer: ```suggestion
	Contributing to the documentation can have many forms:
```

> 	- Correcting spelling or grammatical errors.

Reviewer: ```suggestion
	- Correct spelling or grammatical errors.
```

> 	- Correct incorrect formatting of the docstring. If you see that the official documentation is weirdly displayde or a link is broken, we are very happy if you take some time to correct it.

Reviewer: ```suggestion
	- Correct incorrect formatting of the docstring. If you see that the official documentation is weirdly displayed or a link is broken, we are very happy if you take some time to correct it.
```

	- Correct the shape or dimensions of a docstring input or output tensor.

> 	- Clarify documentation that is hard to understand or incorrect

Reviewer: ```suggestion
	- Clarify documentation that is hard to understand or incorrect.
```

	- Update out-dated code examples.
	- Translating the documentation to another language. See [here](TODO:PVP) for more details.
	- etc...

	Anything displayed on [the official diffusers doc page](TODO:PVP) is part of the official documentation and can be corrected, adjusted in the respective [documentation source](TODO:PVP).

	Please have a look at [this page]( ) on how to verify changes changes made to the documentation locally.
	

### 6. Contribute a [community pipeline](https://github.com/huggingface/diffusers/tree/main/docs/source).

	[Pipelines](TODO:PVP) are usually the first point of contact between the diffusers library and the user.
	Pipelines are examples of how to use diffusers [models](TODO:PVP) and [schedulers](TODO:PVP).
	We support two types of pipelines:

		- Community Pipelines
		- Official Pipelines

> 	Both communtiy and official pipelines follow the same design and consists of the same type of components.

Reviewer: ```suggestion
	Both communtiy and official pipelines follow the same design and consist of the same type of components.
```

> 	Community pipeline are contributed and maintained purely by the **community** and are **not** tested. They 

Reviewer: ```suggestion
	Community pipelines are contributed and maintained purely by the **community** and are **not** tested. They 
```

	reside in [/examples/community]( ) and while they can be accesed via the pypi diffusers package (see [here]( )), their code is not part of the Pypi diffusers package.
	In contrast, official pipelines are tested and maintained by the core diffusers maintainers. Their code
	resides in [src/diffusers/pipelines]( ).

	The reason for the distinction is that the core `diffusers` team simply cannot maintain and test all 
	possible ways diffusion models can be used for inference. 
	Officially released diffusion pipelines, 
	such as [Stable Diffusion]( ) are added to the core src/diffusers/pipelines package which ensures a 
	high quality of maintenance, no backwards-breaking code changes and testing. 

> 	More edge-casy, unofficial pipelines should be added as community pipelines. If usage for a community pipeline is high, they pipeline can be moved to the official pipelines upon request from the community.

Reviewer: ```suggestion
	More edge-casy, unofficial pipelines should be added as community pipelines. If usage for a community pipeline is high, the pipeline can be moved to the official pipelines upon request from the community.
```

    To add a community pipeline, one should add a <name-of-the-community>.py file to [examples/community](TODO:PVP) and adapt the [examples/community/README.md](TODO:PVP) to include an example of the new pipeline.
	
	TODO(PVP:Copy&Paste test from the issue)

	Community pipeline PRs are only checked at a superficial level and it is desired to be maintained by the author.

> 	Contributing a community pipline is a great way to understand how to diffusers models and schedulers work. Having contributed a community pipeline is a great stepping stone to contribute an official pipeline to the 

Reviewer: ```suggestion
	Contributing a community pipeline is a great way to understand how the diffusers models and schedulers work. Having contributed a community pipeline is a great stepping stone to contribute an official pipeline to the 
```
